File: 54321/public/src/games/flipflop.js
import { Cube } from '../core/cube.js';
import { mulberry32, normalizeSeed } from '../core/prng.js';

const TOGGLE_TABLE = [
  [0, 0, 0],
  [1, 2, 3],
  [3, 5, 8],
  [5, 8, 12],
  [8, 16, 32],
];

export class FlipFlop {
  constructor({ dimensions = 2, skillLevel = 0, wrap = true, seed = Date.now() } = {}) {
    this.cube = new Cube();
    this.dimensions = dimensions;
    this.skillLevel = skillLevel;
    this.wrap = wrap;
    this.seed = normalizeSeed(seed);
    this.onCount = 0;
    this.hasWon = false;
    this.actualMoves = 0;
    this.expectedMoves = 0;
    this.scrambleMoves = [];
    this.reset(this.seed);
  }

  reset(seed = Date.now()) {
    this.seed = normalizeSeed(seed);
    const random = mulberry32(this.seed);
    this.cube.fill(0);
    this.onCount = 0;
    this.hasWon = false;
    this.actualMoves = 0;

    const toggles = TOGGLE_TABLE[this.dimensions][this.skillLevel];
    this.expectedMoves = toggles;
    this.scrambleMoves = [];
    const available = Array.from({ length: Cube.ARRAY_LENGTHS[this.dimensions] }, (_, i) => i);

    // Equivalent to the original: choose each scramble cell at most once,
    // then apply a normal Flip-Flop move to it.
    for (let i = 0; i < toggles; i += 1) {
      const pick = Math.floor(random() * available.length);
      const index = available.splice(pick, 1)[0];
      this.scrambleMoves.push(index);
      this.flip(index, { countMove: false });
    }

    this.actualMoves = 0;
    this.hasWon = false;
  }

  configure({ dimensions = this.dimensions, skillLevel = this.skillLevel, wrap = this.wrap } = {}) {
    this.dimensions = dimensions;
    this.skillLevel = skillLevel;
    this.wrap = wrap;
    this.reset(Date.now());
  }

  affectedCells(index) {
    const len = Cube.ARRAY_LENGTHS[this.dimensions];
    if (!Number.isInteger(index) || index < 0 || index >= len) return [];
    const out = [{ index, center: true, axis: null, dimensional: false, wrapped: false }];
    for (const neighbor of Cube.getNeighbors(index, this.dimensions, this.wrap)) {
      out.push({ index: neighbor, center: false, ...Cube.describeNeighbor(index, neighbor, this.dimensions, this.wrap) });
    }
    return out;
  }

  flip(index, { countMove = true } = {}) {
    if (!this.hasWon && countMove) this.actualMoves += 1;
    this.#toggle(index);
    for (const neighbor of Cube.getNeighbors(index, this.dimensions, this.wrap)) {
      this.#toggle(neighbor);
    }
    if (countMove && !this.hasWon && this.onCount === 0) {
      this.hasWon = true;
    }
    return this.hasWon;
  }

  #toggle(index) {
    if (this.cube.toggle(index) === 0) this.onCount -= 1;
    else this.onCount += 1;
  }
}

File: 54321/reference/54321-1.0.2001.11.16/code/flipflop.cpp
    #include <assert.h>
    #include <stdlib.h>
    #include <SDL.h>
    #include "cube.h"
    #include "soundDev.h"
    #include "view.h"
    #include "flipflopView.h"
    #include "flipflop.h"
        NKlein_54321::FlipFlop::FlipFlop(
                Cube* _cube,
                unsigned int _dims,
                unsigned int _skillLevel,
                bool _wrap,
                FlipFlopView* _view
            ) : cube( _cube ),
                dims( _dims ),
                skillLevel( _skillLevel ),
                wrap( _wrap ),
                view( _view )
        {
            assert( cube != 0 );
            assert( dims >= 1 );
            assert( dims <= NKlein_54321::Cube::DIMENSIONS );
            assert( skillLevel < 3 );
            this->reset();
        }
        void
        NKlein_54321::FlipFlop::reset( void )
        {
            *this->cube = 0;
            this->onCount = 0;

                unsigned int table[ Cube::DIMENSIONS+1 ][ 3 ] = {
                    {  0,  0,  0 },
                    {  1,  2,  3 },
                    {  3,  5,  8 },
                    {  5,  8, 12 },
                    {  8, 16, 32 },
                };
            unsigned int toggles = table[ this->dims ][ this->skillLevel ];
            unsigned int len
                = NKlein_54321::Cube::arrayLengths[ this->dims ];

                unsigned int* lut = new unsigned int[ toggles ];
                unsigned int lutLen = 0;

                for ( unsigned int ii=0; ii < toggles; ++ii ) {
                        unsigned int index = random() % len--;
                        for ( unsigned int jj=0; jj < lutLen; ++jj ) {
                            if ( index >= lut[ jj ] ) {
                                ++index;
                            }
                        }

                        this->flip( index, false );

                            unsigned int spot = lutLen;
                            while ( spot > 0 && lut[ spot-1 ] > index ) {
                                lut[ spot ] = lut[ spot-1 ];
                                --spot;
                            }
                            lut[ spot ] = index;
                            ++lutLen;
                }

                delete[] lut;
                this->hasWon = false;
                this->actualMoves = 0;
                this->expectedMoves = toggles;

            if ( this->view != 0 ) {
                this->view->reset();
                this->view->redraw();
            }
        }
        void
        NKlein_54321::FlipFlop::flip(
                unsigned int index, bool update
            )
        {
            unsigned int nn[ 2 * NKlein_54321::Cube::DIMENSIONS ];
            unsigned int nc = this->cube->getNeighbors(
                    nn, index, this->dims, this->wrap
                );

                if ( ! this->hasWon ) {
                    ++this->actualMoves;
                }
                if ( update && this->view != 0 ) {
                    this->view->moveNoise();
                }
                if ( ( (*this->cube)[ index ] ^= 1 ) == 0 ) {
                    --this->onCount;
                } else {
                    ++this->onCount;
                }
                if ( update && this->view != 0 ) {
                    this->view->redraw( index );
                }

            for ( unsigned int ii=0; ii < nc; ++ii ) {
                index = nn[ ii ];
                    if ( ( (*this->cube)[ index ] ^= 1 ) == 0 ) {
                        --this->onCount;
                    } else {
                        ++this->onCount;
                    }
                    if ( update && this->view != 0 ) {
                        this->view->redraw( index );
                    }
            }

                if ( update && ! this->hasWon && this->onCount == 0 ) {
                    this->hasWon = true;
                    if ( this->view != 0 ) {
                        this->view->showWinning(
                                this->actualMoves, this->expectedMoves
                            );
                    }
                }
        }
